### PR TITLE
Support IO_URING transport

### LINKED ISSUE #4163: Support IO_URING transport
I'm implementing support for https://github.com/netty/netty-incubator-transport-io_uring on Vert-x `Transport`, see https://github.com/franz1981/vert.x/tree/4.2.1_iouring for some initial experiment.

### PR DESCRIPTION
**Pull request by franz1981**
Fixes #4163

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -113,6 +114,15 @@
       <artifactId>netty-codec-haproxy</artifactId>
       <optional>true</optional>
     </dependency>
+    <dependency>
+      <groupId>io.netty.incubator</groupId>
+      <artifactId>netty-incubator-transport-native-io_uring</artifactId>
+      <!-- TODO this should be handled on https://github.com/vert-x3/vertx-dependencies -->
```
Comment (franz1981) [inline pom.xml:120]: @vietj what's your suggestion here?

**Comment by franz1981** [comment]
I should add some check re domain socket support on JUnit tests

File: pom.xml (hunk 1 above)
Comment (vietj) [reply pom.xml:120]: yes it should be handled here indeed

File: src/main/java/io/vertx/core/net/impl/ConnectionBase.java [hunk 2]
```diff
@@ -103,6 +107,13 @@ protected ConnectionBase(ContextInternal context, ChannelHandlerContext chctx) {
     closeFuture.onComplete(this::checkCloseHandler);
   }
 
+  public static boolean checkSendFileSupport(final EventLoop eventLoop) {
```
Comment (vietj) [inline src/main/java/io/vertx/core/net/impl/ConnectionBase.java:110]: we need to do that in a way that the IOUring class is not available, have a look at `Transport#nativeTransport` method

**Comment by vietj** [comment]
can you modify the github actions to run io_uring tests ?

File: pom.xml (hunk 1 above)
Comment (franz1981) [reply pom.xml:120]: > yes it should be handled here indeed

you mean https://github.com/vert-x3/vertx-dependencies?
If yes, I suppose I should create a separate linked issue  + PR there too (and vertx parent too)

File: src/main/java/io/vertx/core/net/impl/ConnectionBase.java (hunk 2 above)
Comment (franz1981) [reply src/main/java/io/vertx/core/net/impl/ConnectionBase.java:110]: correct, let me handle CCE

Comment (franz1981) [reply src/main/java/io/vertx/core/net/impl/ConnectionBase.java]: already done

**Comment by JayBenato** [comment]
nice

**Comment by franz1981** [comment]
It seems that `ubuntu-latest` is hitting https://github.com/netty/netty-incubator-transport-io_uring/blob/07bf47cef7436d392623fef34eecf0cd0f9b88bb/README.md#i-tried-to-use-io_uring-but-got-javalangruntimeexception-failed-to-create-io_uring-ring-fd-cannot-allocate-memory

Meaning that:

- we should find a way to change `mlock` limits on it 
- we run it through qemu or docker our tests

For the latter I need some assistance, because I believe it should be beneficial for all vertx tests, not just for these changes, to be a more reproducible build/tests, wdyt @vietj @tsegismont ?

**Comment by franz1981** [comment]
Following up: https://github.community/t/how-to-raise-ulimits-on-the-github-runner/181781

**Comment by franz1981** [comment]
@vietj time for some action :P
https://github.com/netty/netty-incubator-transport-io_uring/pull/159/checks

It seems now we can make use of our CI to run io_uring tests yeeeeeeeeeee

**Comment by franz1981** [comment]
@vietj It seems I cannot make the io_uring part to complete because of intermittent epoll failures :O 
no idea what's going on

**Comment by franz1981** [comment]
This is the list of errors I'm getting by running separately the IO uring tests with the github CI:
```
Tests in error: 
  DatagramTest.testPauseResume:209->AsyncTestBase.await:121->AsyncTestBase.await:133 » IllegalState
  Http1xTest>AsyncTestBase.after:83->HttpTestBase.tearDown:71->VertxTestBase.tearDown:87->AsyncTestBase.close:662->AsyncTestBase.awaitLatch:598->AsyncTestBase.awaitLatch:602->AsyncTestBase.assertTrue:380->AsyncTestBase.handleThrowable:183 » IllegalState
  Http2MetricsTest>AsyncTestBase.after:83->HttpMetricsTestBase.tearDown:57->HttpTestBase.tearDown:71->VertxTestBase.tearDown:87->AsyncTestBase.close:662->AsyncTestBase.awaitLatch:598->AsyncTestBase.awaitLatch:602->AsyncTestBase.assertTrue:380->AsyncTestBase.handleThrowable:183 » IllegalState
  Http2Test>AsyncTestBase.after:83->HttpTestBase.tearDown:71->VertxTestBase.tearDown:87->AsyncTestBase.close:662->AsyncTestBase.awaitLatch:598->AsyncTestBase.awaitLatch:602->AsyncTestBase.assertTrue:380->AsyncTestBase.handleThrowable:183 » IllegalState
io.vertx.core.http.WebSocketTest.testSharedServersRoundRobin(io.vertx.core.http.WebSocketTest)
  Run 1: WebSocketTest.testSharedServersRoundRobin:567->AsyncTestBase.assertTrue:378 null
  Run 2: WebSocketTest>AsyncTestBase.after:83->tearDown:118->VertxTestBase.tearDown:93->AsyncTestBase.tearDown:73->AsyncTestBase.afterAsyncTestBase:163 » IllegalState

io.vertx.core.http.WebSocketTest.testSharedServersRoundRobinWithOtherServerRunningOnDifferentPort(io.vertx.core.http.WebSocketTest)
  Run 1: WebSocketTest.testSharedServersRoundRobinWithOtherServerRunningOnDifferentPort:608->testSharedServersRoundRobin:567->AsyncTestBase.assertTrue:378 null
  Run 2: WebSocketTest>AsyncTestBase.after:83->tearDown:118->VertxTestBase.tearDown:93->AsyncTestBase.tearDown:73->AsyncTestBase.afterAsyncTestBase:163 » IllegalState

  NetTest>AsyncTestBase.after:83->tearDown:132->VertxTestBase.tearDown:87->AsyncTestBase.close:662->AsyncTestBase.awaitLatch:598->AsyncTestBase.awaitLatch:602->AsyncTestBase.assertTrue:380->AsyncTestBase.handleThrowable:183 » IllegalState

Tests run: 5012, Failures: 0, Errors: 7, Skipped: 13
```
Investigating, but still seems a fairly low number (and flaky)

**Comment by franz1981** [comment]
New results after changing the UDP test:
```
Results :

Tests in error: 
  Http1xTest>AsyncTestBase.after:83->HttpTestBase.tearDown:71->VertxTestBase.tearDown:87->AsyncTestBase.close:662->AsyncTestBase.awaitLatch:598->AsyncTestBase.awaitLatch:602->AsyncTestBase.assertTrue:380->AsyncTestBase.handleThrowable:183 » IllegalState
  Http2MetricsTest>AsyncTestBase.after:83->HttpMetricsTestBase.tearDown:57->HttpTestBase.tearDown:71->VertxTestBase.tearDown:87->AsyncTestBase.close:662->AsyncTestBase.awaitLatch:598->AsyncTestBase.awaitLatch:602->AsyncTestBase.assertTrue:380->AsyncTestBase.handleThrowable:183 » IllegalState
  Http2MetricsTest>AsyncTestBase.after:83->HttpMetricsTestBase.tearDown:57->HttpTestBase.tearDown:71->VertxTestBase.tearDown:87->AsyncTestBase.close:662->AsyncTestBase.awaitLatch:598->AsyncTestBase.awaitLatch:602->AsyncTestBase.assertTrue:380->AsyncTestBase.handleThrowable:183 » IllegalState
  Http2Test>AsyncTestBase.after:83->HttpTestBase.tearDown:71->VertxTestBase.tearDown:87->AsyncTestBase.close:662->AsyncTestBase.awaitLatch:598->AsyncTestBase.awaitLatch:602->AsyncTestBase.assertTrue:380->AsyncTestBase.handleThrowable:183 » IllegalState
  NetTest>AsyncTestBase.after:83->tearDown:132->VertxTestBase.tearDown:87->AsyncTestBase.close:662->AsyncTestBase.awaitLatch:598->AsyncTestBase.awaitLatch:602->AsyncTestBase.assertTrue:380->AsyncTestBase.handleThrowable:183 » IllegalState

Tests run: 5012, Failures: 0, Errors: 5, Skipped: 13
```
we're getting there :D

There's some problem while closing the Vert-x loop, checking

**Comment by Sanne** [comment]
Awesome - I'm trying to regularly test Quarkus / Hibernate Reactive on a combination of different options, including io_uring, and if the patch could be included in vert.x it would save me a lot of hassle each time.

Clearly io_uring wasn't super compelling initially, but that's only because we had many more prevalent performance issues to address; now that things are getting really smooth within our libraries, the overhead of those syscalls is getting too nasty - we'll need this :)

**Comment by frank-dspeed** [comment]
LGTM +1 saw this commit and i am Frank1981 👯‍♂️

**Comment by vietj** [comment]
@franz1981 where are we at regarding this PR ?

**Comment by franz1981** [comment]
Hi @vietj 
Here https://github.com/eclipse-vertx/vert.x/pull/4164#issuecomment-1141100555

I am just come back from the summer holidays - will be back on my desk on Monday 👍

**Pull request merged**